app/src/main/java/com/nutomic/syncthingandroid/util/Util.java

```diff
@@ -122,6 +124,81 @@ public static boolean fixAppDataPermissions(Context context) {
         return false;
     }
 
+    /**
+     * Returns if the syncthing binary would be able to write a file into
+     * the given folder given the configured access level.
+     */
+    public static boolean nativeBinaryCanWriteToPath(Context context, String absoluteFolderPath) {
+        final String TOUCH_FILE_NAME = ".stwritetest";
+        Boolean useRoot = false;
+        Boolean prefUseRoot = PreferenceManager.getDefaultSharedPreferences(context)
+            .getBoolean(Constants.PREF_USE_ROOT, false);
+        if (prefUseRoot && Shell.SU.available()) {
+            useRoot = true;
+        }
+
+        // Write permission test file.
+        String touchFile = absoluteFolderPath + "/" + TOUCH_FILE_NAME;
+        int exitCode = runShellCommand("echo \"\" > \"" + touchFile + "\"\n", useRoot);
+        if (exitCode != 0) {
+            String error;
+            switch (exitCode) {
+                case 1:
+                    error = "Permission denied";
+                    break;
+                default:
+                    error = "Shell execution failed";
+            }
+            Log.i(TAG, "Failed to write test file '" + touchFile +
+                "', " + error);
+            return false;
+        }
+
+        // Detected we have write permission.
+        Log.i(TAG, "Successfully wrote test file '" + touchFile + "'");
+
+        // Remove test file.
+        if (runShellCommand("rm \"" + touchFile + "\"\n", useRoot) != 0) {
+            // This is very unlikely to happen, so we have less error handling.
+            Log.i(TAG, "Failed to remove test file");
+        }
+        return true;
+    }
+
+    /**
+     * Run command in a shell and return the exit code.
+     */
+    public static int runShellCommand(String cmd, Boolean useRoot) {
```

Comment:
I'd like to eliminate boilerplate code in a follow-up PR as we have root shell calls spread all over the code and most of them (e.g. niceSyncthing, killSyncthing) can be stripped down to call this new function "runShellCommand".